src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1215,7 +1215,56 @@ ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::limit_init_tm_tx(
     }
     auto term = term_opt.value();
 
+    if (stm->tx_cache_size() > _max_transactions_per_coordinator()) {
+        // lock is sloppy and doesn't guarantee that tx_cache_size
+        // never exceeds _max_transactions_per_coordinator. init_tm_tx
+        // request may pass limit_init_tm_tx but not yet increase
+        // tx_cache_size so there is a small window of time when the
+        // next init tx request may pass too even if the first request
+        // eventually tip tx cache over max transactions per coordinator.
+        // it isn't the problem, the next request will correct it
+        auto init_units = co_await stm->get_tx_thrashing_lock().get_units();
+
+        // similar to double-checked locking pattern
+        // it protects concurrent access to oldest_tx
+        if (stm->tx_cache_size() > _max_transactions_per_coordinator()) {
+            auto tx_opt = stm->oldest_tx();
+            if (!tx_opt) {
+                vlog(
+                  txlog.warn,
+                  "oldest_tx shouldn't return empty when tx cache is at "
+                  "capacity");
+                co_return init_tm_tx_reply{tx_errc::not_coordinator};
+            }
+
+            auto tx = tx_opt.value();
+            vlog(
+              txlog.info,
+              "tx cache is at capacity; expiring oldest tx with id:{}",
+              tx.id);
+            auto tx_units = co_await stm->lock_tx(tx_id, "init_tm_tx");
```

Comment:
i'm curious why the thrashing lock is needed, in that if you observe that the cache size is above the limit, and assume that it'll stay that way, then after you grab this lock, you could recheck?

Reply:
to reduce variability of the code path abd to protect `oldest_tx()` call to avoid contention over the same tx session